rust/consensus/src/lib.rs

```diff
@@ -1,404 +1,290 @@
-            if round_proccessing < leader_round
-                && self.has_path_between_leaders(leader_round, round_proccessing)
-            {
-                leaders_to_commit
-                    .push((self.get_round_leader(round_proccessing), round_proccessing));
-                leader_round = round_proccessing;
+    async fn run(&mut self) {
+        // The consensus state (everything else is immutable).
+        let mut state = State::new(self.genesis.clone());
+
+        // Listen to incoming certificates.
+        while let Some(certificate) = self.rx_primary.recv().await {
+            debug!("Processing {:?}", certificate);
+            let round = certificate.round();
+
+            // Add the new certificate to the local storage.
+            state
+                .dag
+                .entry(round)
+                .or_insert_with(HashMap::new)
+                .insert(certificate.origin(), (certificate.digest(), certificate));
+
+            // Try to order the dag to commit. Start from the highest round for which we have at least
+            // 2f+1 certificates. This is because we need them to reveal the common coin.
+            let r = round - 1;
+
+            // We only elect leaders for even round numbers.
+            if r % 2 != 0 || r < 4 {
+                continue;
             }
-            if round_proccessing > 1 {
-                round_proccessing -= 2;
-            } else {
-                round_proccessing = 0;
-            }
-            debug!("round processing is {:?}", round_proccessing);
-        }
-        leaders_to_commit
-    }
 
-    fn has_path_between_leaders(
-        &self,
-        leader_round: RoundNumber,
-        candidate_round: RoundNumber,
-    ) -> bool {
-        debug!(
-            "Checking path for {:?} and {:?}",
-            leader_round, candidate_round
-        );
-        let causal_blocks =
-            self.get_block_to_commit(self.get_round_leader(leader_round), leader_round);
-        if let Some((candidate_block, _)) = self.get_header_digest_by_node_round(
-            self.get_round_leader(candidate_round),
-            candidate_round,
-        ) {
-            if let Some(causal_blocks_of_round) = causal_blocks.get(&candidate_round) {
-                return causal_blocks_of_round.contains_key(candidate_block);
-            } else {
-                debug!("no causal blocks for round {:?}", candidate_round);
+            // Get the certificate's digest of the leader of round r-2.
+            let leader_round = r - 2;
+            let (leader_digest, leader) = match self.leader(leader_round, &state.dag) {
+                Some(x) => x,
+                None => continue,
+            };
+
+            // If we already ordered this leader, there is nothing to do.
+            let last_committed_round = state.last_committed.values().max().unwrap();
+            if &leader.round() <= last_committed_round {
```

Comment:
why don't you catch this at 113 and wait until here to see if you have already done this round?

Reply:
Fixed